include/ibamr/FESurfaceDistanceEvaluator.h

```diff
@@ -0,0 +1,249 @@
+                               const libMesh::Mesh& mesh,
+                               const libMesh::BoundaryMesh& bdry_mesh,
+                               const int part,
+                               const int gcw = 1,
+                               bool use_extracted_bdry_mesh = true);
+
+    /*!
+     * Destructor for this class.
+     */
+    ~FESurfaceDistanceEvaluator();
+
+    /*!
+     * Map the triangles intersecting a particular grid cell.
+     */
+    void mapIntersections();
+
+    /*!
+     * Get the map maintaining triangle-cell intersection and neighbors.
+     */
+    const std::map<SAMRAI::pdat::CellIndex<NDIM>, std::set<libMesh::Elem*>, IBTK::CellIndexFortranOrder>&
+    getNeighborIntersectionsMap();
+
+    /*!
+     * Compute the signed distance in the viscinity of the finite element mesh.
+     */
+    void computeSignedDistance(int n_idx, int d_idx);
+
+    /*!
+     * Update the sign of the distance function away from the finite element mesh.
+     */
+    static void updateSignAwayFromInterface(int d_idx,
+                                            SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
+                                            double large_distance = s_large_distance);
+
+    /*!
+     * Check whether the grown box and line element intersect in 2D.
+     *
+     *     (box_tl)   *--------------*  (box_tr)
+     *                |              |
+     *                |              |
+     *                |              |
+     *                |              |
+     *                |              |
+     *     (box_bl)   *--------------*  (box_br)
+     */
+    static bool checkIntersection2D(const IBTK::Vector3d& box_bl,
+                                    const IBTK::Vector3d& box_tr,
+                                    const IBTK::Vector3d& box_br,
+                                    const IBTK::Vector3d& box_tl,
+                                    const libMesh::Point& n0,
+                                    const libMesh::Point& n1);
+
+    /*!
+     * Check whether the grown box and triangle element intersect in 3D.
+     *
+     *                  ^   --------------
+     *                  |  |              |                  vert0
+     *  box_half_dx[1]  |  |              |                    *
+     *                  |  |              |                   / \
+     *                  ^  |      *       |                  /   \
+     *                     |   box_center |                 /     \
+     *                     |              |                /       \
+     *                     |              |               /         \
+     *                      --------------               *-----------*
+     *                           <------->
+     *                            box_half_dx[0]      vert1         vert2
+     */
+    static bool checkIntersection3D(const IBTK::Vector3d& box_center,
```

Comment:
@drwells : I hope you liked my ASCII art.

Reply:
It makes the point clearly, so yes :)